diff -ru ../httpd-2.4.12.orig/include/httpd.h ./include/httpd.h
--- ../httpd-2.4.12.orig/include/httpd.h	2014-08-22 20:18:08.000000000 +0200
+++ ./include/httpd.h	2015-06-11 17:05:49.000000000 +0200
@@ -518,6 +518,7 @@
 #define HTTP_UNSUPPORTED_MEDIA_TYPE          415
 #define HTTP_RANGE_NOT_SATISFIABLE           416
 #define HTTP_EXPECTATION_FAILED              417
+#define HTTP_MISDIRECTED_REQUEST             421
 #define HTTP_UNPROCESSABLE_ENTITY            422
 #define HTTP_LOCKED                          423
 #define HTTP_FAILED_DEPENDENCY               424
diff -ru ../httpd-2.4.12.orig/modules/http/http_protocol.c ./modules/http/http_protocol.c
--- ../httpd-2.4.12.orig/modules/http/http_protocol.c	2014-10-16 23:42:45.000000000 +0200
+++ ./modules/http/http_protocol.c	2015-06-11 17:04:49.000000000 +0200
@@ -135,7 +135,7 @@
     NULL, /* 418 */
     NULL, /* 419 */
     NULL, /* 420 */
-    NULL, /* 421 */
+    "421 Misdirected Request",
     "422 Unprocessable Entity",
     "423 Locked",
     "424 Failed Dependency",
@@ -1296,6 +1296,9 @@
     case HTTP_NETWORK_AUTHENTICATION_REQUIRED:
         return("<p>The client needs to authenticate to gain\n"
                "network access.</p>\n");
+    case HTTP_MISDIRECTED_REQUEST:
+        return("<p>The client needs to use a new connection for this \n"
+                   "request as it does not match the SNI name used.</p>\n");
     default:                    /* HTTP_INTERNAL_SERVER_ERROR */
         /*
          * This comparison to expose error-notes could be modified to
diff -ru ../httpd-2.4.12.orig/modules/ssl/ssl_engine_kernel.c ./modules/ssl/ssl_engine_kernel.c
--- ../httpd-2.4.12.orig/modules/ssl/ssl_engine_kernel.c	2015-06-10 11:46:00.000000000 +0200
+++ ./modules/ssl/ssl_engine_kernel.c	2015-06-11 16:58:34.000000000 +0200
@@ -203,7 +203,7 @@
                 ap_log_error(APLOG_MARK, APLOG_ERR, 0, r->server, APLOGNO(02032)
                             "Hostname %s provided via SNI and hostname %s provided"
                             " via HTTP are different", servername, host);
-                return HTTP_BAD_REQUEST;
+                return HTTP_MISDIRECTED_REQUEST;
             }
         }
         else if (((sc->strict_sni_vhost_check == SSL_ENABLED_TRUE)
